src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/EnvelopedCms.cs

```diff
@@ -0,0 +1,268 @@
+
+            if (_decryptorPal != null)
+            {
+                _decryptorPal.Dispose();
+                _decryptorPal = null;
+            }
+
+            int version;
+            ContentInfo contentInfo;
+            AlgorithmIdentifier contentEncryptionAlgorithm;
+            X509Certificate2Collection originatorCerts;
+            CryptographicAttributeObjectCollection unprotectedAttributes;
+            _decryptorPal = PkcsPal.Instance.Decode(encodedMessage, out version, out contentInfo, out contentEncryptionAlgorithm, out originatorCerts, out unprotectedAttributes);
+            Version = version;
+            ContentInfo = contentInfo;
+            ContentEncryptionAlgorithm = contentEncryptionAlgorithm;
+            Certificates = originatorCerts;
+            UnprotectedAttributes = unprotectedAttributes;
+
+            // Desktop compat: Encode() after a Decode() returns you the same thing that ContentInfo.Content does.
+            _encodedMessage = contentInfo.Content.CloneByteArray();
+
+            _lastCall = LastCall.Decode;
+        }
+
+        //
+        // Decrypt() overloads: Recipients invoke Decrypt() after a successful Decode(). Afterwards, invoke ContentInto to get the decrypted content.
+        //
+        public void Decrypt()
+        {
+            DecryptContent(RecipientInfos, null);
+        }
+
+        public void Decrypt(RecipientInfo recipientInfo)
+        {
+            if (recipientInfo == null)
+                throw new ArgumentNullException(nameof(recipientInfo));
+
+            DecryptContent(new RecipientInfoCollection(recipientInfo), null);
+        }
+
+        public void Decrypt(RecipientInfo recipientInfo, X509Certificate2Collection extraStore)
+        {
+            if (recipientInfo == null)
+                throw new ArgumentNullException(nameof(recipientInfo));
+
+            if (extraStore == null)
+                throw new ArgumentNullException(nameof(extraStore));
+
+            DecryptContent(new RecipientInfoCollection(recipientInfo), extraStore);
+        }
+
+        public void Decrypt(X509Certificate2Collection extraStore)
+        {
+            if (extraStore == null)
+                throw new ArgumentNullException(nameof(extraStore));
+
+            DecryptContent(RecipientInfos, extraStore);
+        }
+
+        private void DecryptContent(RecipientInfoCollection recipientInfos, X509Certificate2Collection extraStore)
+        {
+            switch (_lastCall)
+            {
+                case LastCall.Ctor:
+                    throw new InvalidOperationException(SR.Cryptography_Cms_MessageNotEncrypted);
+
+                case LastCall.Encrypt:
+                    throw PkcsPal.Instance.CreateDecryptAfterEncryptException();
+
+                case LastCall.Decrypt:
+                    throw PkcsPal.Instance.CreateDecryptTwiceException();
+
+                case LastCall.Decode:
+                    break; // This is the expected state. Fall through.
```

Comment:
Usually I think of "fall through" to mean "go to the next case label".  I'd just remove the second sentence, but it's not a big deal.